proposals/inline-classes.md

```diff
@@ -0,0 +1,163 @@
+# Inline classes
+
+* **Type**: Design proposal
+* **Author**: Mikhail Zarechenskiy
+* **Contributors**: Andrey Breslav, Denis Zharkov, Roman Elizarov, Stanislav Erokhin
+* **Status**: Under consideration
+* **Prototype**: Implemented in Kotlin 1.2.30
+
+Discussion of this proposal is held in [this issue](https://github.com/Kotlin/KEEP/issues/104).
+
+## Summary
+
+Currently, there is no performant way to create wrapper for a value of a corresponding type. The only way is to create a usual class, 
+but the use of such classes would require additional heap allocations, which can be critical for many use cases.    
+
+We propose to support identityless inline classes that would allow to introduce wrappers for a values without additional overhead related 
+to additional heap allocations.
+
+## Motivation / use cases
+
+Inline classes allow to create wrappers for a value of a certain type and such wrappers would be fully inlined. 
+This is similar to type aliases but inline classes are not assignment-compatible with the corresponding underlying types.
+
+Use cases:
+
+- Unsigned types
+```
+inline class UInt(private val value: Int) { ... }
+inline class UShort(private val value: Short) { ... }
+inline class UByte(private val value: Byte) { ... }
+inline class ULong(private val value: Long) { ... }
+```
+
+- Native types like `size_t` for Kotlin/Native
+- Inline enum classes 
+    - Int enum for [Android IntDef](https://developer.android.com/reference/android/support/annotation/IntDef.html)
+    - String enum for Kotlin/JS (see [WebIDL enums](https://www.w3.org/TR/WebIDL-1/#idl-enums))
+- Units of measurement
+- Result type (aka Try monad) [KT-18608](https://youtrack.jetbrains.com/issue/KT-18608)
+- Inline property delegates
+- Inline wrappers
+
+## Description
+
+Inline classes are declared using soft keyword `inline` and must have a single property:
+```
+inline class Foo(val i: Int)
+```
+Property `i` defines type of the underlying runtime representation for inline class `Foo`, while at compile time type will be `Foo`.
+
+From language point of view, inline classes can be considered as restricted classes, they can declare various members, operators, 
+have generics. 
+
+Example:
+```
+inline class Name(val s: String) : Comparable<Name> {
+    override fun compareTo(other: Name): Int = s.compareTo(other.s)
+    
+    fun greet() {
+        println("Hello, $s")
+    }
+}    
+
+fun greet() {
+    val name = Name("Kotlin") // there is no actual instantiation of class `Name`
+    name.greet() // method `greet` is called as a static method
+}
+```
+
+## Current limitations
+
+Currently, inline classes must satisfy the following requirements:
+
+- Inline class must have a public primary constructor with a single value parameter
+- Inline class must have a single read-only (`val`) property as an underlying value, which is defined in primary constructor
```

Comment:
Why is it impossible to use `var`? Is it technical limitation or design decision?

Reply:
I'd say it's both :)
Short answer is that because inline classes are identityless, they don't have state

Reply:
```kotlin
inline class Name(var s: String)

val name = Name("Alan")
name.s = "Bob"
```

Is `name` read-only or not?
How estimate the impact of change the `Name.s` property in a shared boxed type?

Reply:
Any problems with single-property class which violates other statements?
```
inline class Something(t: T) {
    init {
        // some init, checks
    }
    val value: U = transform(t)
}
```

Reply:
@Miha-x64 This is the same as if we have private primary constructor + public secondary constructor, therefore we are going to prohibit such cases for now